Merge multiple settlements in driver

vkgnosis opened the pull request:
Finally enabled merging of multiple settlements.

How do you think we should test this? I could adapt the e2e test so that we assert that in a single run it settles 4 orders instead of the current two.

vkgnosis committed:
Merge multiple settlements in driver

nlordell commented on solver/src/driver.rs:
```diff
@@ -314,14 +337,24 @@ struct RatedSettlement {
+}
+
 #[cfg(test)]
 mod tests {
```
Might be worth adding some unit tests for the new logic.

nlordell reviewed (commented):
Changes look good.

I would like to see a couple of unit tests that demonstrates how merging affects the `SolverWIthSettlements` result, but it may be quite verbose to setup. Maybe if more of the code in the `filter_map` closure were in a function, it might be easier to test?

fleupold commented on solver/src/driver.rs:
```diff
@@ -314,14 +337,24 @@ struct RatedSettlement {
+        .next()
+        .ok_or_else(|| anyhow!("no settlements"))?;
+    for settlement in settlements.take(MAX_MERGED_SETTLEMENTS - 1) {
+        merged = merged.merge(settlement)?;
```
If a specific settlement cannot be merged into the existing bunch, shouldn't we still try to merge the rest of the settlements? This way we can still reduce the number of total transactions we are waiting for.

vkgnosis replied:
Fixed.

fleupold commented on solver/src/driver.rs:
```diff
@@ -23,6 +23,8 @@ use std::{
 // protect against a gas estimator giving bogus results that would drain all our funds.
 const GAS_PRICE_CAP: f64 = 500e9;
 
+const MAX_MERGED_SETTLEMENTS: usize = 5;
```
Should this be configurable via command line so it's easier to adapt?

fleupold reviewed (approved):
Second the suggestion to add unit tests. The driver logic is becoming more and more complex and doesn't yet have good coverage.

Otherwise, great PR 🎉 !

vkgnosis commented:
I added two more commits which:
* move some of the settlement handling logic into another file
* split out merge related function and added unit test
* make max merge count configurable

nlordell commented on solver/src/driver.rs:
```diff
@@ -204,25 +175,31 @@ impl Driver {
+            .filter_map(solver_settlements::filter_bad_settlements)
+            .inspect(|(name, settlements)| {
+                for settlement in settlements.iter() {
+                    tracing::debug!("solver {} found solution:\n {:?}", name, settlement,);
```
ubernit: trailing comma
```suggestion
                    tracing::debug!("solver {} found solution:\n {:?}", name, settlement);
```

nlordell commented on solver/src/driver/solver_settlements.rs:
```diff
@@ -0,0 +1,204 @@
+
+// Return None if the result is an error or there are no settlements remaining after removing
+// settlements with no trades.
+pub fn filter_bad_settlements(
```
"bad" - so judgemental :stuck_out_tongue:.

nlordell reviewed (approved):
Looks amazing
